### PR TITLE
[stable9] try connecting to redis 3 times

### PR DESCRIPTION
**Pull request by butonic**
On long uploads the connection to redis might timeout. OC should automatically reconnect. This PR adds a retry mechanism.

### CODE REVIEW

**Comment by butonic** [comment]
@DeepDiver1975 @PVince81 or should we retry with all cache implementations?

Does the
- [ ] Memcached implementation throw exceptions?
- [ ] APCu implementation throw exceptions?

**Comment by butonic** [comment]
needs forward port but we are testing this at a 9.0.4 instance

**Comment by butonic** [comment]
@DeepDiver1975 jenkins bulid broken?
```
23:55:43 PHP Fatal error:  Uncaught Error: Class 'Phar' not found in /usr/local/bin/phpunit:12
```

File: lib/private/memcache/redis.php [hunk 1]
```diff
@@ -185,24 +191,39 @@ public function cas($key, $old, $new) {
...
 	}
+
+	public function retryOp($name, array $params = []) {
+		$tries = 0;
+		$ex = null;
+		do {
+			try {
+				return call_user_func_array([self::$cache, $name], $params);
+			} catch (\RedisException $ex) {
+				self::$cache->close();
```
Comment (DeepDiver1975) [inline lib/private/memcache/redis.php]: could close throw an exception?

**Comment by DeepDiver1975** [comment]
> On long uploads the connection to redis might timeout.

isn't this something where we better advise people to increase the timeout?
is this a client or an server config value within redis? or both?

**Comment by butonic** [comment]
both. Actually you would want to know when your infrastructure has hiccups so you should set the redis timeout to st. like 3 sec but allow for multiple retries. redis itself also has a timeout. You might want to use haproxy in front of redis. haproxy kills connections after 5min. radis after 30min I think. You don't want your processes to keep connections for that long.

If an upload takes 30min (as in too long) oc should be able to reconnect to the cache. Otherwise tho upload will fail. Uncreasing the connection timeout to 30min is just a workaround.

**Comment by butonic** [comment]
OH and with redis every connection requires an open file descriptor on the server, something that you may quickly run out of ...

**Comment by DeepDiver1975** [comment]
> something that you may quickly run out of ...

should not really be of an issue ....




> If an upload takes 30min (as in too long) oc should be able to reconnect to the cache.

indeed

File: lib/private/memcache/redis.php (hunk 1 above)
Comment (butonic) [inline lib/private/memcache/redis.php]: No, throwing an exception is done with `zend_throw_exception`, eg: https://github.com/phpredis/phpredis/blob/5bff86db67aa4b8559d9df93b722618ba62db1c2/library.c#L1785 I don't see that in close or the functions it calls:

- [close](https://github.com/phpredis/phpredis/blob/develop/redis.c#L831)
- [redis_sock_get_connected](https://github.com/phpredis/phpredis/blob/fa3267933a6ad72b0359f12f861789192491c36f/redis.c#L486)
- [redis_sock_disconnect](https://github.com/phpredis/phpredis/blob/5bff86db67aa4b8559d9df93b722618ba62db1c2/library.c#L1684)

**Comment by butonic** [comment]
@DeepDiver1975
```
10:34:58 [PR-26012] Running shell script
10:34:59 + export NOCOVERAGE=1
10:34:59 + unset USEDOCKER
10:34:59 + phpenv local 7.0
10:34:59 + ./autotest.sh sqlite
10:34:59 Using PHP executable /opt/phpenv/shims/php
10:34:59 rbenv: php: command not found
10:34:59 
10:34:59 The `php' command exists in these PHP versions:
10:34:59   5.4
10:34:59   5.4.45
10:34:59   5.5
10:34:59   5.5.35
10:34:59   5.6
10:34:59   5.6.21
10:34:59   5.6.24
10:34:59   5.6.25
10:34:59   7.0.10
10:34:59   7.0.8
10:34:59   7.1
10:34:59   7.1.0alpha3
10:34:59   7.1.0beta1
10:34:59   7.1.0RC1
10:34:59 
10:34:59 ./autotest.sh: line 72: [: : integer expression expected
10:34:59 phpunit version >= 4.4 required. Version found: 
```

**Comment by PVince81** [comment]
Rebased.

@butonic will this need forward ports ?

@DeepDiver1975 do you agree with the PR ?

**Comment by butonic** [comment]
@michaelstingl 00005779 can you get us the phpredis version and check with the script in https://github.com/phpredis/phpredis/issues/76#issuecomment-58809399

**Comment by butonic** [comment]
For completeness sake:

> We’re connecting to Redis through HAproxy. When connecting directly to Redis, this issue is gone.
In HAproxy, the default timeout for connections was 5 minutes. We’ve increased this number to 30 minutes.
>
> I’m not a developer, but I’m guessing Owncloud opens a Redis connection once. During that connection, a large file is downloaded from Owncloud to the client, while the Redis connections remains idle.
When the download takes longer than 5 minutes (the Redis timeout), the Redis connection is gone.

@michaelstingl the redis verision can be obtained from the php info

**Comment by DeepDiver1975** [comment]
Can we please get the version of the php redis module as used 

> We’re connecting to Redis through HAproxy. When connecting directly to Redis, this issue is gone.

Missconfiguration?

**Comment by butonic** [comment]
https://github.com/phpredis/phpredis/releases/tag/2.2.6 seems to contain the fix

**Comment by butonic** [comment]
```
php -i tells me:
Redis Support => enabled
Redis Version => 2.2.7

Haproxy server (Centos6):
haproxy-1.5.14

Redis server (Centos6):
redis-3.0.7-1
```

**Comment by butonic** [comment]
should be handled by phpredis itself.

**Pull request closed without merging**